### PR TITLE
ARXIVNG-1994: AutoTeX Log: highlight lexicon terms in autotex.log

### PR DESCRIPTION
**Pull request by DavidLFielding**
This PR implements 'basic' highlighting in the autotex log. Ok, so maybe not SO basic any more.

ARXIVNG-1994: AutoTeX Log: highlight lexicon terms in autotex.log

The following highlighting features are supported:

    - Analyzes TeX runs and ignores hypertex runs when normal TeX/LaTeX/PDFLaTeX runs exist
    - Creates simple summary of TeX runs
    - Creates simple display of highlighting key
    - match a term and highlight the matching text (basic highlighting)
    - apply filters at specific run levels
         Examples: 
             Only apply a filter on the last TeX run (References undefined)
             Apply the same filter on different runs specifying different classes.

Right now the filters are basic in the sense we could associate a 'subroutine' that does more in terms of extended markup, extracting errors summary, etc.

Notes about the current state of code:

    I've included examples of all of the classes that mark of the first few lines of every autotex log.

    Example filter on PATH highlights first run as 'help' class and later runs as 'suggestion' class to show that you are able to change the highlighting of the same error in later runs.

    You are also able to 'ignore' errors and warnings. This is intended to indicate we detect the error/warning but determine that it is not relevant or important in the current run.

    Example filters:

    # Examples
        ['suggestion', r':.*PATH.*', 'second'], # Note ORDER is critical here
        ['help', r':.*PATH.*', 'first'], # otherwise this rule trumps all PATH rules
        ['suggestion', r'Set working directory to.*', ''],
        ['ignore', 'Setting unix time to current time.*', ''],
        ['ignore','Using source archive.*',''],
        ['info', r'Using directory .* for processing.', ''],
        ['warning', r'Copied file .* into working directory.', ''],
        ['danger', 'nostamp: will not stamp PostScript', ''],
        ['danger', r'TeX/AutoTeX.pm', ''],
        ['fatal', r'override', ''],

    # Real example where we skip undefined references until later runs:
    # This shows undefined Citations are highlighted starting in the first run and 
    # undefined References are highlighted starting in second run. These will both
    # be set to 'last' for production.

        # Warnings
        ['warning', r'Citation.*undefined', 'first'], # needs to be 'last'
        ['warning', r'Reference.*undefined', 'second'], # needs to be 'last'

Recent Enhancements:

Added abort error for XeTeX/LuaTeX. When abort condition is detected all highlighting is disabled except for abort conditions.

Jake wanted me to display the submission_id when we generate the XeTeX/LuaTeX errors so I figured out how to pass additional arguments to filter script. Also passing submission status.

Added documentation and examples for regex filter expressions.

Added new processing summary that indicates status of compilation (submission).

If processing 'succeeded' we now examine the last run to see if there are warnings/error messages that might impact the PDF. We indicate if we found warnings/errors in the summary at top of log.

I'm done with this now.

### CODE REVIEW

File: submit/filters/tex_filters.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+                new_log = new_log + f"Skip line {line}\n"
+                break
+
+        if done_with_line:
+            continue
+
+
+
+        # Ignore, Info, Help, Warning, Danger, Fatal
+        for level, filter, run in filters:
```
Comment (erickpeirson) [inline submit/filters/tex_filters.py]: ``filter()`` is a built-in; suggest renaming

File: submit/filters/tex_filters.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+ENABLE_TEX = r'(\~+\sRunning tex.*\s\~+)'
+ENABLE_LATEX = r'(\~+\sRunning latex.*\s\~+)'
+ENABLE_PDFLATEX = r'(\~+\sRunning pdflatex.*\s\~+)'
+
+DISABLE_HTEX = r'(\~+\sRunning htex.*\s\~+)'
+DISABLE_HLATEX = r'(\~+\sRunning hlatex.*\s\~+)'
+DISABLE_HPDFLATEX = r'(\~+\sRunning hpdflatex.*\s\~+)'
+
+RUN_ORDER = ['last', 'first', 'second', 'third', 'forth']
```
Comment (erickpeirson) [inline submit/filters/tex_filters.py]:
```suggestion
RUN_ORDER = ['last', 'first', 'second', 'third', 'fourth']
```

File: submit/filters/tex_filters.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+def compilation_log_display(autotex_log: str) -> str:
```
Comment (erickpeirson) [inline submit/filters/tex_filters.py]: For testing, I would focus on examples. Here is a similar situation: https://github.com/arXiv/arxiv-base/blob/master/arxiv/base/urls/tests/test_urlize.py

So if you have a sample line for each of the patterns you are testing, you would be asserting that the expected HTML-annotated output is produced

File: submit/filters/tex_filters.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    """
+
+    :param autotex_log:
```
[the remark this answers is not in the document]
Comment (DavidLFielding) [reply submit/filters/tex_filters.py]: Darn. I forgot to fix this after I set up new project. Fixed.

File: submit/filters/tex_filters.py (hunk 2 above)
Comment (DavidLFielding) [reply submit/filters/tex_filters.py]: ugh! fixed.

File: submit/filters/tex_filters.py (hunk 3 above)
Comment (DavidLFielding) [reply submit/filters/tex_filters.py]: So basically test at the routine level, pass in a controlled log and assert the desired outcome. I will try to add something shortly.

Comment (erickpeirson) [reply submit/filters/tex_filters.py]: Yep; and if possible using short examples -- easier to see what is being tested

**Pull request merged**